perses/analysis/free_energies.py

```diff
@@ -0,0 +1,242 @@
+        errs = []
+
+        if len(f_works) > 10 and len(r_works) > 10:
+            for _ in range(n_bootstrap):
+                f = random.choices(f_works, k=len(f_works))
+                r = random.choices(r_works, k=len(r_works))
+                fe, err = BAR(np.asarray(f), np.asarray(r))
+                fes.append(fe)
+                errs.append(err)
+
+        return fes, errs, f_works, r_works
+
+    if cache_dir is not None:
+        # cache results in local directory
+        memory = joblib.Memory(backend="local", location=cache_dir, verbose=0)
+        bootstrap_BAR = memory.cache(_bootstrap_BAR)
+    else:
+        bootstrap_BAR = _bootstrap_BAR
+
+    for d in tqdm.tqdm(details.values()):
+        RUN = d["directory"]
+        if show_plots:
+            fig, axes = plt.subplots(ncols=2, nrows=1, figsize=(10, 5))
+        for i, phase in enumerate(projects.keys()):
+            if show_plots:
+                axes[i].set_title(phase)
+
+            all_forward = []
+            all_reverse = []
+            # There will be 6 gens for this project I think
+            for gen_id in range(0, 7):
+                fes, errs, f_works, r_works = bootstrap_BAR(RUN, phase, gen_id)
+                d[f"{phase}_fes_GEN{gen_id}"] = fes
+                d[f"{phase}_dfes_GEN{gen_id}"] = errs
+                all_forward.extend(f_works)
+                all_reverse.extend(r_works)
+            #         print(all_forward)
+            if len(all_forward) < 10:
+                print(f"Cant calculate {RUN} {phase}")
+                continue
+            if len(all_reverse) < 10:
+                print(f"Cant calculate {RUN} {phase}")
+                continue
+            if show_plots:
+                sns.kdeplot(all_forward, shade=True, color="cornflowerblue", ax=axes[i])
+                sns.rugplot(
+                    all_forward,
+                    ax=axes[i],
+                    color="cornflowerblue",
+                    alpha=0.5,
+                    label=f"forward : N={len(f_works)}",
+                )
+                sns.rugplot(
+                    all_forward,
+                    ax=axes[i],
+                    color="darkblue",
+                    label=f"forward (gen0) : N={len(f_works)}",
+                )
+                sns.rugplot(
+                    [-x for x in all_reverse],
+                    ax=axes[i],
+                    color="mediumvioletred",
+                    label=f"reverse (gen0) : N={len(r_works)}",
+                )
+                sns.kdeplot(
+                    [-x for x in all_reverse], shade=True, color="hotpink", ax=axes[i]
+                )
+                sns.rugplot(
+                    [-x for x in all_reverse],
+                    ax=axes[i],
+                    color="hotpink",
+                    alpha=0.5,
+                    label=f"reverse : N={len(r_works)}",
+                )
+
+            if any([True for x in [all_reverse, all_forward] if len(x) < 20]):
```

Comment:
`any` supports generators so it can short-circuit as soon as a True value is needed.

```suggestion
            if any(len(x) < 20 for x in (all_reverse, all_forward)):
```

Reply:
Thanks; looks like the check was redundant anyway. Removed in 6d0f121.